packages/fxa-settings/src/models/integrations/base-integration.ts

```diff
@@ -55,18 +47,174 @@ export interface IntegrationFeatures {
+  subscriptionProductId: string;
+  subscriptionProductName: string;
+}
+
+export interface RelierClientInfo {
+  clientId: string | undefined;
+  imageUri: string | undefined;
+  serviceName: string | undefined;
+  redirectUri: string | undefined;
+  trusted: boolean | undefined;
+}
+
+/**
+ * A convenience interface for easy mocking / testing. Use this type in components.
  */
+// export interface RelierData {
+//   name: string;
+//   context: string | undefined;
+//   email: string | undefined;
+//   entrypoint: string | undefined;
+//   entrypointExperiment: string | undefined;
+//   entrypointVariation: string | undefined;
+//   resetPasswordConfirm: boolean | undefined;
+//   setting: string | undefined;
+//   service: string | undefined;
+//   style: string | undefined;
+//   uid: string | undefined;
+//   utmCampaign: string | undefined;
+//   utmContent: string | undefined;
+//   utmMedium: string | undefined;
+//   utmSource: string | undefined;
+//   utmTerm: string | undefined;
+
+//   clientInfo: Promise<RelierClientInfo> | undefined;
+//   subscriptionInfo: Promise<RelierSubscriptionInfo> | undefined;
+// }
+
+export interface ResumeTokenInfo {
+  // TBD
+}
+
+// export interface Relier extends RelierUrlData {
+//   getServiceName(): Promise<string>;
+//   getClientInfo(): Promise<RelierClientInfo | undefined>;
+//   isOAuth(): boolean;
+//   isSync(): Promise<boolean>;
+//   shouldOfferToSync(view: string): boolean;
+//   wantsKeys(): boolean;
+//   wantsTwoStepAuthentication(): boolean;
+//   isTrusted(): boolean;
+//   validate(): void;
+//   getService(): string | undefined;
+//   getRedirectUri(): string | undefined;
+// }
+
+export interface RelierAccount {
+  uid: string;
+  email: string;
+  sessionToken: string;
+  verifyIdToken(
+    idTokenHint: string,
+    clientId: string,
+    gracePeriod: number
+  ): Promise<{ sub: string }>;
+  isDefault(): unknown;
+}
+
+export abstract class Integration<
+  T extends IntegrationFeatures = IntegrationFeatures
+> {
+  type: IntegrationType;
+  public features: T = {} as T;
+  // TODO fix data type
+  public data: any;
+  /** Lazy loaded client info. */
+  clientInfo: Promise<RelierClientInfo> | undefined;
+  /** Lazy loaded subscription info. */
+  subscriptionInfo: Promise<RelierSubscriptionInfo> | undefined;
+
+  // TODO fix data type
+  constructor(type: IntegrationType, data: any) {
+    this.type = type;
+    this.data = data;
+  }
+
+  protected setFeatures(features: Partial<T>) {
+    this.features = { ...this.features, ...features } as T;
+  }
+
+  // TODO: do we need this/isSync?
+  isOAuth(): boolean {
+    return false;
+  }
+
+  async isSync(): Promise<boolean> {
+    return false;
+  }
+
+  async getClientInfo(): Promise<RelierClientInfo | undefined> {
+    return undefined;
+  }
+
+  async getServiceName(): Promise<string> {
+    // If the service is not defined, then check the client info
+    if (!!this.data.service) {
+      if (this.clientInfo) {
```

Comment:
Should be addressed in FXA-6836 👍

Reply:
The intent is that if the service null or undefined or empty, then we fallback to the client info. I found this all a bit confusing as well, and really wish there was just a single way to determine this. IMO, we should always just query for client info, but I am not entirely sure that this will _always_ produce backwards compatible behavior. The fix in #15606, essentially made is so that clientInfo was queried and used, and this cleared up the reported testing issues.

As for `null` vs `undefined`, keep in mind that this all gets bound at runtime with the `@bind` decorator, and there is some casting going when getting the values out of the URL (since everything is held as a string, this unavoidable). Anyways, it's quite possible something in the binding code is returning null even though the type is `string | undefined`. I'll revisit this when I look into FXA-8099.

Reply:
@dschom Ah, so, for some reason I thought your PR had changed it from `!` to `!!`. It was the opposite, which makes a lot more sense. 😆